session/session.go

```diff
@@ -564,6 +565,16 @@ func (s *session) doCommit(ctx context.Context) error {
 
 func (s *session) commitTxnWithTemporaryData(ctx context.Context, txn kv.Transaction) error {
 	sessVars := s.sessionVars
+
+	// Note: ignore auto-commit transactions because auto-commit transactions don't write untouched index mutations
+	// into the mem buffer, which invalidates the check.
+	// Only applies to user DML transactions.
+	if sessVars.ConnectionID != 0 && sessVars.EnableMutationChecker && sessVars.TxnCtx.IsExplicit {
+		if err := tables.CheckTxnConsistency(txn); err != nil {
```

Comment:
I think maybe you can rename this function to show that it checks based on the counts of entries. Otherwise I'm afraid it might looks kinda confusing.